source_code/main_mcu/src/LOGIC/logic_encryption.c

```diff
@@ -252,4 +270,194 @@ void logic_encryption_ctr_decrypt(uint8_t* data, uint8_t* cred_ctr, uint16_t dat
+{
+    br_hmac_update(&logic_encryption_hmac_ctx, data, len);
+}
+
+/*! \fn     logic_encryption_sha256_hmac_final(uint8_t * hmac_out)
+*   \brief  Finalize the HMAC and output the value to the argument
+*   \param  hash  The final HMAC value (output)
+*/
+void logic_encryption_sha256_hmac_final(uint8_t * hmac_out)
+{
+    br_hmac_out(&logic_encryption_hmac_ctx, hmac_out);
+}
+
+/*! \fn     logic_encryption_ecc256_init(void)
+*   \brief  Initialize the ECC256 crypto object
+*/
+void logic_encryption_ecc256_init(void)
+{
+    #define SEED_LEN 8
+    uint8_t seed[SEED_LEN];
+
+    logic_encryption_br_ec_algo = &br_ec_p256_m15;
+    logic_encryption_br_ec_algo_id = BR_EC_secp256r1;
+
+    rng_fill_array(seed, SEED_LEN);
+    br_hmac_drbg_init(&logic_encryption_hmac_drbg_ctx, &br_sha256_vtable, seed, SEED_LEN);
+}
+
+/*! \fn     logic_encryption_ecc256_sign(uint8_t const *data, int len, uint8_t * sig)
+*   \brief  Cryptographically sign input data and return the signature in arg.
+*   \param  data    data to sign
+*   \param  len     length of data to sign
+*   \param  sig     output signature
+*/
+void logic_encryption_ecc256_sign(uint8_t const *data, int len, uint8_t * sig)
+{
+    size_t result = br_ecdsa_i15_sign_raw(logic_encryption_br_ec_algo, logic_encryption_sha256_ctx.vtable, data, &logic_encryption_signing_key, sig);
+
+    if (result == 0)
+    {
+        //TODO: Log(TAG_ERR, "error, signing failed!\n");
+        while(1);
+    }
+    //Clear private key used to limit leaking
+    memset(logic_encryption_priv_key_buf, 0, sizeof(logic_encryption_priv_key_buf));
+}
+
+/*! \fn     _create_br_key(uint8_t const *priv_key, br_ec_private_key *br_key)
+*   \brief  Create key compatible with BearSSL library
+*   \param  priv_key    Buffer containing the private key
+*   \param  br_key      Output BearSSL compatible key
+*/
+static void _create_br_key(uint8_t const *priv_key, br_ec_private_key *br_key)
+{
+    memcpy(logic_encryption_priv_key_buf, priv_key, sizeof(logic_encryption_priv_key_buf));
+
+    br_key->curve = logic_encryption_br_ec_algo_id;
+    br_key->xlen = FIDO2_PRIV_KEY_LEN;
+    br_key->x = logic_encryption_priv_key_buf;
+}
+
+/*! \fn     logic_encryption_ecc256_load_key(uint8_t const *key)
+*   \brief  Load a key from a buffer
+*   \param  key     The key to load
+*/
+void logic_encryption_ecc256_load_key(uint8_t const *key)
+{
+    _create_br_key(key, &logic_encryption_signing_key);
+}
+
+/*! \fn     logic_encryption_ecc256_generate_private_key(uint8_t *priv_key)
+*   \brief  Generate a private key using ECC256
+*   \param  priv_key    Output private key
+*   \return size of key
+*/
+size_t logic_encryption_ecc256_generate_private_key(uint8_t *priv_key)
+{
+    size_t result;
+
+    result = br_ec_keygen(&logic_encryption_hmac_drbg_ctx.vtable, logic_encryption_br_ec_algo, NULL, priv_key, logic_encryption_br_ec_algo_id);
+    if (result == 0)
+    {
+        //Log("error, generating private key failed\n");
+        while(1);
```

Comment:
what could cause this failure? is it so unlikely that we want to freeze the device over this?

Reply:
I'll look into changing it to return error instead if there is a good way to "recover" from this error.

Reply:
up to you.... I'm fine with the while() as long as the conditions that lead there are known

Reply:
Programmer error would lead us to this.